src/components/periods/PeriodSelect.js

```diff
@@ -16,124 +17,127 @@ import { getYear } from '../../util/time.js'
-        className: PropTypes.string,
-        errorText: PropTypes.string,
-        period: PropTypes.shape({
-            id: PropTypes.string.isRequired,
-            startDate: PropTypes.string,
-        }),
-        periodType: PropTypes.string,
-    }
-
-    state = {
-        year: null,
-        periods: null,
-    }
-
-    componentDidMount() {
-        this.setPeriods()
-    }
-
-    componentDidUpdate(prevProps, prevState) {
-        const { periodType, period, onChange } = this.props
-        const { year, periods } = this.state
-
-        if (periodType !== prevProps.periodType) {
-            this.setPeriods()
-        } else if (periods && !period) {
-            onChange(filterFuturePeriods(periods)[0] || periods[0]) // Autoselect most recent period
-        }
-
-        // Change period if year is changed (but keep period index)
-        if (period && prevState.periods && year !== prevState.year) {
-            const periodIndex = prevState.periods.findIndex(
-                (item) => item.id === period.id
-            )
-            onChange(periods[periodIndex])
-        }
-    }
-
-    render() {
-        const { periodType, period, onChange, className, errorText } =
-            this.props
-        const { periods } = this.state
-
-        if (!periods) {
-            return null
+const PeriodSelect = ({
+    onChange,
+    className,
+    errorText,
+    firstDate,
+    lastDate,
+    period,
+    periodType,
+}) => {
+    const [year, setYear] = useState(getYear(period?.startDate || lastDate))
+    const prevYear = usePrevious(year)
+
+    // Set periods when periodType or year changes
+    /* eslint-disable react-hooks/exhaustive-deps */
+    const periods = useMemo(
+        () =>
+            periodType
+                ? getFixedPeriodsByType({
+                      periodType,
+                      year,
+                      firstDate,
+                      lastDate,
+                  })
+                : [period], // saved map period (not included in depency array by design)
+        [periodType, year, firstDate, lastDate]
+    )
+    /* eslint-enable react-hooks/exhaustive-deps */
+
+    const periodIndex = useMemo(
+        () => period && periods.findIndex((p) => p.id === period.id),
+        [period, periods]
+    )
+
+    const prevPeriodIndex = usePrevious(periodIndex)
+
+    // Increment/decrement year
+    const changeYear = useCallback(
+        (change) => {
+            const newYear = year + change
+
+            if (
+                (!firstDate || newYear >= getYear(firstDate)) &&
+                (!lastDate || newYear <= getYear(lastDate))
+            ) {
+                setYear(newYear)
+            }
+        },
+        [year, firstDate, lastDate]
+    )
+
+    // Autoselect most recent period
+    useEffect(() => {
+        if (!period && periods) {
+            onChange(filterFuturePeriods(periods)[0] || periods[0])
         }
+    }, [period, periods, year, onChange])
```

Comment:
`year` is not used in the effect. If you still want it in the dependency list, maybe add a comment

Reply:
Fixed.